### ISSUE TITLE
`watch` command not working 

### ISSUE THREAD

**Issue by maxxst** [Ceasar/staticjinja#56]
So I tried to use the `watch` command and it immediately crashes with following error:

```
Traceback (most recent call last):
  File "/Users/max/Envs/staticjinja/bin/staticjinja", line 11, in <module>
    sys.exit(main())
  File "/Users/max/Envs/staticjinja/lib/python3.4/site-packages/staticjinja/cli.py", line 66, in main
    site.render(use_reloader=use_reloader)
  File "/Users/max/Envs/staticjinja/lib/python3.4/site-packages/staticjinja/staticjinja.py", line 296, in render
    Reloader(self).watch()
  File "/Users/max/Envs/staticjinja/lib/python3.4/site-packages/staticjinja/reloader.py", line 52, in watch
    import easywatch
  File "/Users/max/Envs/staticjinja/lib/python3.4/site-packages/easywatch/__init__.py", line 1, in <module>
    from easywatch import watch
ImportError: cannot import name 'watch'
```

I'm using a clean virtualenv with just staticjinja installed

I hope you can help me

**Comment by Ceasar** [comment]
Weird. Looking into it..

For now, can you try typing the following in a terminal and pasting the results?

```
>>> import easywatch
>>> dir(easywatch)
>>> help(easywaytch)
```

**Comment by Ceasar** [comment]
Works for me on a fresh install.

```
$ virtualenv
$ source env/bin/activate
$ pip install staticjinja
$ mkdir templates
$ staticjinja watch
```

How did you install it?

**Comment by maxxst** [comment]
I used virtualenvwrapper ...

so I tried importing easywatch and there seems to be the issue already

```
Python 3.4.3 (default, Jul 13 2015, 12:18:23)
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import easywatch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/max/Envs/staticjinja/lib/python3.4/site-packages/easywatch/__init__.py", line 1, in <module>
    from easywatch import watch
ImportError: cannot import name 'watch'
```

**Comment by Ceasar** [comment]
Right... The `__init__` file makes that import. Can you open `/Users/max/Envs/staticjinja/lib/python3.4/site-packages/easywatch/__init__.py` and find where `watch` is supposed to be?

**Comment by maxxst** [comment]
somehow it loaded the "old" version saying

```
from easywatch import watch
```

I changed it locally to
```
from .easywatch import watch
```

now it works ..

**Comment by Ceasar** [comment]
Strange. I'm going to close this for now since I can't reproduce it, but will keep an eye out for this issue in the future.

**Comment by Ceasar** [comment]
Ran into something similar today. Seems like the issue is that I set up an old version of my other repository "easywatch" which has the bug you mentioned. Fixing this now..

**Comment by mentat-fr** [comment]
Hi Ceaser,

I've exactly same error:

       (staticjinja)  staticjinja watch --outpath=output
        Rendering fr/index.html...
        Rendering index.html...
        Watching '/http/templates' for changes...
        Press Ctrl+C to stop.
        Traceback (most recent call last):
          File "/virtualenvs/staticjinja/bin/staticjinja", line 11, in <module>
            sys.exit(main())
          File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/cli.py", line 66, in main
            site.render(use_reloader=use_reloader)
          File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/staticjinja.py", line 296, in render
            Reloader(self).watch()
          File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/reloader.py", line 52, in watch
            import easywatch
          File "/virtualenvs/staticjinja/lib/python3.4/site-packages/easywatch/__init__.py", line 1, in <module>
            from easywatch import watch
        ImportError: cannot import name 'watch'


My install is fresh and up to date: 

    python 3.4
    easywatch-0.0.3
    statiticjinja-0.1

Python debug:

    >>> import easywatch
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/easywatch/__init__.py", line 1, in <module>
        from easywatch import watch
    ImportError: cannot import name 'watch'

If I modify `easywatch/__init__.py` and place dot:

    from .easywatch import watch

Python is now ok:

    >>> import easywatch
    >>> dir(easywatch)
    ['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'easywatch', 'watch']
 
Then watch again, I obtain:

    (staticjinja)  staticjinja watch --outpath=output
    Rendering fr/index.html...
    Rendering index.html...
    Watching '/http/templates' for changes...
    Press Ctrl+C to stop.
    created /http/templates/4913
    modified /http/templates/4913
    Exception in thread Thread-1:
    Traceback (most recent call last):
      File "/usr/lib64/python3.4/threading.py", line 920, in _bootstrap_inner
        self.run()
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/watchdog/observers/api.py", line 199, in run
        self.dispatch_events(self.event_queue, self.timeout)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/watchdog/observers/api.py", line 368, in dispatch_events
        handler.dispatch(event)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/watchdog/events.py", line 322, in dispatch
        self.on_any_event(event)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/easywatch/easywatch.py", line 25, in wrapper
        return handler(event.event_type, event.src_path)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/reloader.py", line 47, in event_handler
        templates = self.site.get_dependencies(filename)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/staticjinja.py", line 278, in get_dependencies
        return [self.get_template(filename)]
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/staticjinja/staticjinja.py", line 94, in get_template
        return self._env.get_template(template_name)
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/jinja2/environment.py", line 812, in get_template
        return self._load_template(name, self.make_globals(globals))
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/jinja2/environment.py", line 774, in _load_template
        cache_key = self.loader.get_source(self, name)[1]
      File "/virtualenvs/staticjinja/lib/python3.4/site-packages/jinja2/loaders.py", line 187, in get_source
        raise TemplateNotFound(template)
    jinja2.exceptions.TemplateNotFound: 4913

I can build my pages with no problem.
Thanks.
F.

**Comment by PatrickMassot** [comment]
It  seems that pypi still has the broken version. Changing "from easywatch import watch"
to "from .easywatch import watch" works for me.

**Comment by mentat-fr** [comment]
Thank you Patrick, I'll try and check it cause I use only the build command.

**Comment by Ceasar** [comment]
Hm, I thought I had fixed this. I'll investigate it this evening.

**Comment by PatrickMassot** [comment]
Did you reach any conclusion on this? This issue is preventing me to tox my smart dependency refactor for staticjinja because tox keeps downloading the broken easywatch from pip.

Note that upstream staticjinja doesn't have this tox issue only because of its test suite lacks full coverage (especially for cli.py).

**Comment by PatrickMassot** [comment]
Ah ah, I'm no longer stuck! I can add to tox.ini:

```python
[testenv]
whitelist_externals=sed
commands=
	sed 's/from easywatch import watch/from .easywatch import watch/' -i {envsitepackagesdir}/easywatch/__init__.py
	py.test tests
```

This is not something we want to keep there forever though...

**Comment by Ceasar** [comment]
Yeah that's bad. I unfortunately got sidetracked. Probably makes sense to remove the dependency on easywatch and just write that part ourselves (or copy easywatch). I only wrote easywatch because I thought it might be useful to others but it seems like it's more trouble than it's worth. I'll put it on my to-do list and shoot for getting it done by this weekend.

**Comment by PatrickMassot** [comment]
I'll probably copy easywatch for now and we'll see after deciding on what to do with staticjinja. My work on staticjinja is taking longer than I anticipated but I probably got carried away. 

Tonight I spend the whole time failing to understand how I could write a functional test that actually uses watchdog, does some modification to file but doesn't wait for Ctrl-C to stop watchdog after those modifications.

**Comment by Ceasar** [comment]
Take your time. :) No rush here, would rather get things right.

And yeah, that's going to be tricky. You should probably look at the tests for watchdog for help. The way I'd test it though it to simulate the events rather than actually starting watchdog. In imperative languages, the way we do that is via mocking. And specifically in Python it's via the `mock` library, https://pypi.python.org/pypi/mock, which is in the standard lib for Python 3.

I actually find that library hard to understand, so if you'd prefer you can delegate writing the tests to me. If you prefer to do it yourself though, I recommend checking out these two articles which helped me solve some "gotchas": http://alexmarandon.com/articles/python_mock_gotchas/ and http://www.voidspace.org.uk/python/mock/patch.html#where-to-patch

**Comment by PatrickMassot** [comment]
Thanks for your advice. I didn't think of looking at tests for watchdog itself. I sort of know about mock and I know there are traps. The situation is worse than what is described in your links. The patch decorator doesn't play nicely with py.test fixtures so it cannot be used when fixtures are used. There is also a pytest-mock plugin and the monkeypatch standard py.test fixture. They all behave slightly differently (I think). I use a semi-deterministic approach to using mocks with lots of trial and errors. And when it seems to work I always make sure I can get the test to fail by changing a couple of values because sometimes the test passes simply because a mock swallows everything as explained [here](http://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html).

Anyway, that's not what was my problem here. I tried the uncoupled road and, as expected, it makes it much easier to write unit tests with mocks and test double everywhere. But it also become very easy to write a fully covering and passing unit test suite for a software that simply doesn't work. That's why I wanted functional tests. I partially gave up on the reloader. My test uses a different watch function that still calls watchdog (in a separate thread to be able to modify files while watching) but doesn't have the loop waiting for Ctrl-C (instead it calls ``wait`` on a ``threading.Event`` controlled by the test function). The issue is not mocking, the problem is that Ctrl-C exception goes to a random thread and I got crazy.

Then I wasted a huge amount of time trying to find a reliable way to compare the content of two directories to check the output of the functional tests. It's partly because ``filecmp.dircmp`` doesn't want to actually look at the content. But it was also because by default Jinja2 doesn't end rendered template with EOL. It took me ages to understand why opening a rendered template in vim and saving it was enough to change it. Both things together created much confusion.

Anyway, I hope I'll have a new PR before the week-end.
